tests/test_options.py

```diff
@@ -1100,188 +1232,293 @@ def test_type_from_flag_value():
+        ({"type": bool, "default": False, "flag_value": False}, ["--foo"], False),
+        ({"type": bool, "default": None, "flag_value": True}, ["--foo"], True),
+        ({"type": bool, "default": None, "flag_value": False}, ["--foo"], False),
+        ({"type": str, "default": True, "flag_value": True}, ["--foo"], "True"),
+        ({"type": str, "default": True, "flag_value": False}, ["--foo"], "False"),
+        ({"type": str, "default": False, "flag_value": True}, ["--foo"], "True"),
+        ({"type": str, "default": False, "flag_value": False}, ["--foo"], "False"),
+        ({"type": str, "default": "foo", "flag_value": True}, ["--foo"], "True"),
+        ({"type": str, "default": "foo", "flag_value": False}, ["--foo"], "False"),
+        ({"type": str, "default": "foo", "flag_value": "bar"}, ["--foo"], "bar"),
+        ({"type": str, "default": "foo", "flag_value": None}, ["--foo"], None),
+        ({"type": str, "default": None, "flag_value": True}, ["--foo"], "True"),
+        ({"type": str, "default": None, "flag_value": False}, ["--foo"], "False"),
     ],
 )
-def test_flag_value_and_default(runner, opt_params, pass_flag, expected):
+def test_flag_value_and_default(runner, opt_params, args, expected):
     @click.command()
     @click.option("--foo", is_flag=True, **opt_params)
     def cmd(foo):
-        click.echo(repr(foo))
+        click.echo(repr(foo), nl=False)
 
-    result = runner.invoke(cmd, ["--foo"] if pass_flag else [])
-    assert result.output == f"{expected!r}\n"
+    result = runner.invoke(cmd, args)
+    assert result.output == repr(expected)
 
 
 @pytest.mark.parametrize(
-    "opts",
+    ("args", "opts"),
     [
-        {"type": bool, "default": "foo"},
-        {"type": bool, "flag_value": "foo"},
-        {"type": click.BOOL, "default": "foo"},
-        {"type": click.BOOL, "flag_value": "foo"},
+        ([], {"type": bool, "default": "foo"}),
+        ([], {"type": click.BOOL, "default": "foo"}),
+        (["--foo"], {"type": bool, "flag_value": "foo"}),
+        (["--foo"], {"type": click.BOOL, "flag_value": "foo"}),
     ],
 )
-def test_invalid_flag_definition(runner, opts):
+def test_invalid_flag_definition(runner, args, opts):
     @click.command()
     @click.option("--foo", is_flag=True, **opts)
     def cmd(foo):
         click.echo(foo)
 
-    result = runner.invoke(cmd, ["--foo"])
+    result = runner.invoke(cmd, args)
     assert (
         "Error: Invalid value for '--foo': 'foo' is not a valid boolean"
         in result.output
     )
 
 
+@pytest.mark.parametrize(
+    ("default", "args", "expected"),
+    [
+        (None, [], "A real None"),
+        (UNSET, [], "A real UNSET"),
+        ("random", [], "random"),
+        (None, ["--js"], "js"),
+        (None, ["--xml"], "xml"),
+    ],
+)
+def test_flag_value_callback(runner, default, args, expected):
+    """
+    Reproduction of the issue reported in
+    https://github.com/pallets/click/pull/3030#discussion_r2271571819
```

Comment:
Describe what the test does in the docstring, the link should only be extra context.